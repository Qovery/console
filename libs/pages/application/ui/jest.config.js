module.exports = {
  displayName: 'pages-application-ui',
  preset: '../../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../../coverage/libs/pages/application/ui',
}
