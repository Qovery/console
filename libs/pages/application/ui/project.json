{
  "sourceRoot": "libs/pages/application/ui/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/pages/application/ui/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/pages/application/ui"],
      "options": {
        "jestConfig": "libs/pages/application/ui/jest.config.ts",
        "passWithNoTests": true
      }
    }
  }
}
