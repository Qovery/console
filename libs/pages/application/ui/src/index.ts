export * from './lib/components/deployments-page/deployments-page'
export * from './lib/components/about/about'
export * from './lib/components/general/general'
export * from './lib/components/container/container'
