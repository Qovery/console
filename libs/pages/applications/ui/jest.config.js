module.exports = {
  displayName: 'pages-applications-ui',
  preset: '../../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../../coverage/libs/pages/applications/ui',
}
