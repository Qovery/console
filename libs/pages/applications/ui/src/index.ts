export * from './lib/components/table-row-applications/table-row-applications'
export * from './lib/components/general/general'
export * from './lib/components/container/container'
