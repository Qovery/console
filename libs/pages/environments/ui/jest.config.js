module.exports = {
  displayName: 'pages-environments-ui',
  preset: '../../../../jest.preset.js',
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../../../coverage/libs/pages/environments/ui',
}
