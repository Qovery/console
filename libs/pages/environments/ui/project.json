{
  "root": "libs/pages/environments/ui",
  "sourceRoot": "libs/pages/environments/ui/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/pages/environments/ui/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/pages/environments/ui"],
      "options": {
        "jestConfig": "libs/pages/environments/ui/jest.config.js",
        "passWithNoTests": true
      }
    }
  }
}
