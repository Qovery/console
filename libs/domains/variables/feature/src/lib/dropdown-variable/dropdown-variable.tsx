import * as DropdownMenu from '@radix-ui/react-dropdown-menu'
import { type PropsWithChildren, useState } from 'react'
import { Icon, InputSearch, Popover, Tooltip, Truncate, dropdownMenuItemVariants } from '@qovery/shared/ui'
import { twMerge } from '@qovery/shared/util-js'
import { useVariables } from '../hooks/use-variables/use-variables'

export interface DropdownVariableProps extends PropsWithChildren {
  environmentId: string
  onChange: (value: string) => void
}

export function DropdownVariable({ environmentId, onChange, children }: DropdownVariableProps) {
  const { data: variables = [] } = useVariables({
    parentId: environmentId,
    scope: 'ENVIRONMENT',
  })
  const [open, setOpen] = useState(false)

  const [searchTerm, setSearchTerm] = useState('')
  const filteredVariables = variables.filter((variable) =>
    variable.key.toLowerCase().includes(searchTerm.toLowerCase())
  )

  // XXX: https://github.com/radix-ui/primitives/issues/1342
  // We are waiting for radix combobox primitives
  // So we are using DropdownMenu.Root in combination of Popover.Root
  // to get the flexibility of Popover.Root but keeping the accessiblity of
  // DropdownMenu.Root for entries.
  // So both open state should be sync
  return (
    <DropdownMenu.Root open={open} onOpenChange={(open) => setOpen(open)}>
      <Popover.Root open={open} onOpenChange={(open) => setOpen(open)}>
        <Popover.Trigger>{children}</Popover.Trigger>
        <DropdownMenu.Content asChild>
          <Popover.Content className="flex max-h-60 w-[248px] min-w-[248px] flex-col p-2">
            {/* 
                `stopPropagation` is used to prevent the event from `DropdownMenu.Root` parent
                fix issue with item focus if we use input search
                https://github.com/radix-ui/primitives/issues/2193#issuecomment-1790564604 
              */}
            <div className="bg-white dark:bg-neutral-700" onKeyDown={(e) => e.stopPropagation()}>
              <InputSearch
                placeholder="Search..."
                className="mb-1"
                onChange={(value) => setSearchTerm(value)}
                autofocus
              />
            </div>
            <div className="overflow-y-auto">
              {filteredVariables.length > 0 ? (
                filteredVariables.map((variable) => (
                  <Popover.Close key={variable.key} onClick={() => setSearchTerm('')}>
                    <DropdownMenu.Item
                      className={twMerge(
                        dropdownMenuItemVariants({ color: 'brand' }),
                        'flex h-[52px] items-center justify-between gap-1 px-2 py-1.5'
                      )}
                      onClick={() => onChange(variable.key)}
                    >
                      <div className="flex flex-col items-start justify-center gap-1">
                        <span className="text-sm font-medium">
                          <Truncate text={variable.key} truncateLimit={21} />
                        </span>

                        {variable.service_name ? (
                          <span className="truncate text-xs font-normal">
                            <Truncate text={variable.service_name} truncateLimit={30} />
                          </span>
                        ) : (
                          <span className="text-xs font-normal text-neutral-300">no service</span>
                        )}
                      </div>
                      {variable.description && (
                        <Tooltip content={variable.description} side="bottom">
                          <span>
                            <Icon iconName="info-circle" iconStyle="regular" className="text-neutral-400" />
                          </span>
                        </Tooltip>
                      )}
                    </DropdownMenu.Item>
                  </Popover.Close>
                ))
              ) : (
                <div className="px-3 py-6 text-center">
                  <Icon iconName="wave-pulse" className="text-neutral-350" />
                  <p className="mt-1 text-xs font-medium text-neutral-350">No result for this search</p>
                </div>
              )}
            </div>
          </Popover.Content>
        </DropdownMenu.Content>
      </Popover.Root>
    </DropdownMenu.Root>
  )
}

export default DropdownVariable
