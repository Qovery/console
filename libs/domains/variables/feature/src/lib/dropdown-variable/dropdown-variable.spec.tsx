import { Button } from '@qovery/shared/ui'
import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import { DropdownVariable, type DropdownVariableProps } from './dropdown-variable'

const props: DropdownVariableProps = {
  environmentId: '000',
  onChange: jest.fn(),
}

jest.mock('../hooks/use-variables/use-variables', () => ({
  useVariables: () => ({
    data: [
      {
        id: 'c0277184-1fe0-4f17-b09a-58eee2c05701',
        created_at: '2023-10-25T09:13:34.723567Z',
        updated_at: '2023-10-25T09:13:34.723568Z',
        key: 'QOVERY_CONTAINER_Z04308DE2_HOST_INTERNAL',
        value: 'app-z04308de2-back-end',
        mount_path: null,
        scope: 'BUILT_IN',
        overridden_variable: null,
        aliased_variable: null,
        variable_type: 'BUILT_IN',
        service_id: '04308de2-af27-405f-9e95-570fa94ed577',
        service_name: 'back-end-A',
        service_type: 'CONTAINER',
        owned_by: 'QOVERY',
        description: 'test',
        is_secret: false,
      },
    ],
  }),
}))

describe('DropdownVariable', () => {
  it('should render successfully', () => {
    const { baseElement } = renderWithProviders(
      <DropdownVariable {...props}>
        <Button>Button</Button>
      </DropdownVariable>
    )
    expect(baseElement).toBeTruthy()
  })

  it('should render dropdown content when clicked', async () => {
    const { userEvent } = renderWithProviders(
      <DropdownVariable {...props}>
        <Button>Button</Button>
      </DropdownVariable>
    )

    const button = screen.getByRole('button')
    await userEvent.click(button)

    const dropdownContent = await screen.findByRole('menu')

    expect(dropdownContent).toMatchSnapshot()
  })
})
