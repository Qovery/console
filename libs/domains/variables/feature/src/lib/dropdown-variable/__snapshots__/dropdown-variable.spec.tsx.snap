// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownVariable should render dropdown content when clicked 1`] = `
<div
  aria-labelledby="radix-:rf:"
  aria-orientation="vertical"
  class="rounded bg-white shadow-xl outline-none data-[state=open]:data-[side=bottom]:animate-slidein-up-sm-faded data-[state=open]:data-[side=left]:animate-slidein-right-sm-faded data-[state=open]:data-[side=right]:animate-slidein-left-sm-faded data-[state=open]:data-[side=top]:animate-slidein-down-sm-faded dark:border dark:border-neutral-500 dark:bg-neutral-550 flex max-h-60 w-[248px] min-w-[248px] flex-col p-2"
  data-align="center"
  data-orientation="vertical"
  data-radix-menu-content=""
  data-side="bottom"
  data-state="open"
  dir="ltr"
  id="radix-:rg:"
  role="menu"
  style="outline: none; --radix-dropdown-menu-content-transform-origin: var(--radix-popper-transform-origin); --radix-dropdown-menu-content-available-width: var(--radix-popper-available-width); --radix-dropdown-menu-content-available-height: var(--radix-popper-available-height); --radix-dropdown-menu-trigger-width: var(--radix-popper-anchor-width); --radix-dropdown-menu-trigger-height: var(--radix-popper-anchor-height); --radix-popover-content-transform-origin: var(--radix-popper-transform-origin); --radix-popover-content-available-width: var(--radix-popper-available-width); --radix-popover-content-available-height: var(--radix-popper-available-height); --radix-popover-trigger-width: var(--radix-popper-anchor-width); --radix-popover-trigger-height: var(--radix-popper-anchor-height); pointer-events: auto;"
  tabindex="-1"
>
  <div
    class="bg-white dark:bg-neutral-700"
  >
    <div
      class="relative w-full mb-1"
    >
      <span
        class="icon-solid-magnifying-glass shrink-0 absolute left-3 top-1/2 block -translate-y-1/2 text-base leading-none text-neutral-400 dark:text-white"
        role="img"
      />
      <input
        class="w-full rounded border border-neutral-250 pl-10 pr-6 text-neutral-400 placeholder:text-neutral-350 focus:border-brand-400 focus:outline-none focus:transition-[border-color] dark:border-neutral-350 dark:bg-transparent dark:text-white dark:placeholder:text-neutral-250 h-9 text-xs"
        data-testid="input-search"
        name="search"
        placeholder="Search..."
        type="text"
      />
    </div>
  </div>
  <div
    class="overflow-y-auto"
  >
    <div
      class="text-sm font-medium rounded-sm outline-none data-[highlighted]:bg-brand-50 data-[highlighted]:text-brand-500 dark:data-[highlighted]:bg-neutral-400 dark:text-neutral-100 cursor-pointer text-neutral-400 hover:bg-brand-50 hover:text-brand-500 dark:hover:bg-neutral-550 flex h-[52px] items-center justify-between gap-1 px-2 py-1.5"
      data-orientation="vertical"
      data-radix-collection-item=""
      role="menuitem"
      tabindex="-1"
      type="button"
    >
      <div
        class="flex flex-col items-start justify-center gap-1"
      >
        <span
          class="text-sm font-medium"
        >
          <span
            class=""
            data-state="closed"
            data-testid="truncate-text"
          >
            QOVERY_CONTAINER_Z043
            ...
          </span>
        </span>
        <span
          class="truncate text-xs font-normal"
        >
          back-end-A
        </span>
      </div>
      <span
        class=""
        data-state="closed"
      >
        <i
          aria-hidden="true"
          class="fa-regular fa-info-circle text-neutral-400"
        />
      </span>
    </div>
  </div>
</div>
`;
